torchdet3d/utils/tracking_tools.py

```diff
@@ -0,0 +1,308 @@
+        self.id = ID
+        self.boxes = [bbox]
+        self.kps = [kps]
+        self.timestamps = [time]
+        self.no_updated_frames = 0
+        self.align_kp = align_kp
+
+    def get_end_time(self):
+        return self.timestamps[-1]
+
+    def get_start_time(self):
+        return self.timestamps[0]
+
+    def get_last_box(self):
+        return self.boxes[-1]
+
+    def get_last_kp(self):
+        return self.kps[-1]
+
+    def __len__(self):
+        return len(self.timestamps)
+
+    def _interpolate(self, target_box, target_kp, timestamp, skip_size):
+        last_box = self.get_last_box()
+        last_kp = self.get_last_kp()
+        for t in range(1, skip_size):
+            interp_box = [int(b1 + (b2 - b1) / skip_size * t) for b1, b2 in zip(last_box, target_box)]
+            interp_kp = [k1 + (k2 - k1) / skip_size * t for k1, k2 in zip(last_kp, target_kp)]
+            self.boxes.append(interp_box)
+            self.kps.append(interp_kp)
+            self.timestamps.append(self.get_end_time() + 1)
+
+    def _filter_last_3d_box(self, filter_speed, add_treshold, no_updated_frames_treshold):
+        if self.timestamps[-1] - self.timestamps[-2] == 1:
+            num_keypoints = len(self.kps[-2]) // 2
+            self.kps[-2] = np.array(self.kps[-2]).reshape(num_keypoints,2)
+            self.kps[-1] = self.kps[-1].reshape(num_keypoints,2)
+            # compute average distance before run
+            add_dist = np.mean(np.linalg.norm(self.kps[-1] - self.kps[-2], axis=1))
+            if self.align_kp:
+                indexes_to_revert = self._align_kp_positions()
+                rearranged_kps = self.kps[-1][indexes_to_revert]
+                add_dist_after = np.mean(np.linalg.norm(rearranged_kps - self.kps[-2], axis=1))
+                if add_dist_after < add_dist:
+                    considered_kps = rearranged_kps
+                    add_dist = add_dist_after
+                else:
+                    considered_kps = self.kps[-1]
+            else:
+                considered_kps = self.kps[-1]
+            # if add distance is appropriate for previous frame by given treshold
+            # then we smooth kps with EMA
+            if add_dist < add_treshold:
+                self.no_updated_frames = 0
+                filtered_kps = (1 - filter_speed) * self.kps[-2] + filter_speed * considered_kps
+            elif self.no_updated_frames > no_updated_frames_treshold:
+                # if bbox haven't been updated too long -> interrupt EMA
+                # and get new bbox
+                filtered_kps = considered_kps
+            else:
+                # if not -> use bbox from previous frame
+                filtered_kps = self.kps[-2]
+                self.no_updated_frames += 1
+
+            self.kps[-1] = tuple(filtered_kps.reshape(-1).tolist())
+
+    def _align_kp_positions(self):
+        # store indexes for matching
+        indexes = dict(zip(range(9), range(9)))
```

Comment:
Would `list(range(self.kps[-1].shape[0]))` be more suitable? Seems like keys of the dict are not supposed to be changed.